modules/partition-replicator/src/test/java/org/apache/ignite/internal/partition/replicator/PartitionReplicaLifecycleManagerTest.java

```diff
@@ -324,4 +378,55 @@ void producesEventsOnPartitionDestroy() {
         assertThat(beforeReplicaDestroyedFuture.thenApply(LocalPartitionReplicaEventParameters::zonePartitionId), willBe(zonePartitionId));
         assertThat(afterReplicaDestroyedFuture.thenApply(LocalPartitionReplicaEventParameters::zonePartitionId), willBe(zonePartitionId));
     }
+
+    @Test
+    public void partitionLifecycleManagerStopsCorrectWhenAllComponentsAreStoppedFine() throws Exception {
+        assertDoesNotThrow(() -> partitionReplicaLifecycleManager.beforeNodeStop());
+
+        assertThat(partitionReplicaLifecycleManager.stopAsync(), willCompleteSuccessfully());
+
+        verify(replicaManager, times(CatalogUtils.DEFAULT_PARTITION_COUNT)).stopReplica(any());
+    }
+
+    @Test
+    public void partitionLifecycleManagerStopsCorrectWhenReplicasAreStoppedExceptionally() throws Exception {
+        when(replicaManager.stopReplica(any())).thenThrow(new NodeStoppingException());
+
+        assertDoesNotThrow(() -> partitionReplicaLifecycleManager.beforeNodeStop());
+
+        assertThat(partitionReplicaLifecycleManager.stopAsync(), willCompleteSuccessfully());
+
+        // One more because of exceptional mock above counts.
+        verify(replicaManager, times(CatalogUtils.DEFAULT_PARTITION_COUNT + 1)).stopReplica(any());
+
+        // Do reset for correct replica manager stop on tear down.
+        reset(replicaManager);
+    }
+
+    @Test
+    public void partitionLifecycleManagerStopsCorrectWhenTxStatePartitionStoragesAreStoppedExceptionally() throws Exception {
+        doReturn(commonZonePartitionResources).when(zoneResourcesManager).getZonePartitionResources(any());
+
+        int defaultZoneId = 0;
+        List<ZonePartitionResources> defaultZoneResources = IntStream.range(0, CatalogUtils.DEFAULT_PARTITION_COUNT)
+                .mapToObj(partId -> new ZonePartitionId(defaultZoneId, partId))
+                .map(partitionReplicaLifecycleManager::zonePartitionResources)
+                .collect(Collectors.toList());
+
+        defaultZoneResources.forEach(resources -> {
+            TxStatePartitionStorage txStorage = resources.txStatePartitionStorage();
+
+            doThrow(new RuntimeException()).when(txStorage).close();
```

Comment:
So, you need all these machinations with the TX storages only to throw an exception on close? `ZoneResourcesManager` is already a spy here, can you simply throw an exception from its `close` method?

Reply:
I tired to save `TableManager`'s tests logic, that is why the storages' `close` mocks with exception throwing.

Reply:
Yes, it's a bit different but it tests a close situation. I'm against your current approach, because we have to do changes in production code to simulate this situation.

Reply:
Drilling holes by introducing unnecessary protected methods does not make the code more testable and should always be the last resort

Reply:
1. I don't see really difference between protected methods for testing and `@TestOnly` getters in `IgniteImpl`.
2. I don't see better solution to test storages closing issues. One of the problems is the in-place instantiation and this method also improves it. Is there any better solution?

Reply:
Made the create method public, check the correctness in `ThreadAssertingTxStateStorage` please.